### PR TITLE
Add permission requester

### PR DESCRIPTION
**Pull request by zsoltk**

### CODE REVIEW

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Interactor.kt [hunk 1]
```diff
@@ -56,11 +59,18 @@ abstract class Interactor<C : Parcelable, V : RibView>(
     val isAttached: Boolean
         get() = behaviorRelay.value == ACTIVE
 
+    internal var tag = "${this::class.java.name}.${UUID.randomUUID()}"
```
Comment (ShikaSD) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Interactor.kt]: Probably, we should reuse the tag from the node to be consistent

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+import com.badoo.ribs.core.requestcode.RequestCodeRegistry
+import com.jakewharton.rxrelay2.PublishRelay
+import io.reactivex.Observable
+
+class PermissionRequesterImpl(
+    private val activity: AppCompatActivity,
+    private val requestCodeRegistry: RequestCodeRegistry
+) : PermissionRequester, PermissionRequestResultHandler {
+
+    private val events = HashMap<Int, HashMap<Int, PublishRelay<RequestPermissionsEvent>>>()
```
Comment (ShikaSD) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: I would suggest to have one relay with filter instead of hashmap of hashmaps.
Then we can just wrap the RequestPermissionEvent with data class holding it, requestCode and groupId. I feel like this should be a bit simpler than this construct.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Interactor.kt (hunk 1 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Interactor.kt]: I tried that first, but there's a problem: if you need to setup a subscription in your `Interactor`, then it's `id()` method will be invoked, and since `Node` is lateinit (will be created after `Interactor`), it crashes if it tries to forward the call. So `Interactor` needs to identify itself regardless of `Node`.

My idea is that we can remove `tag` from `Node` altogether.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 2 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: I'm waiting for everyone's opinion on this. So far there's 1 for, 1 against the current impl, and I don't have strong feelings about it.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Interactor.kt (hunk 1 above)
Comment (ShikaSD) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Interactor.kt]: Alright, probably makes sense

Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Interactor.kt]: Also an idea for future PR to unify identifiers (there's one field called `identifier` in `Node` that extends `Rib` interface, there's `tag`, theres `ribID`). So probably `identifier: Rib` could be extracted and injected both to `Node` and to `Interactor`, and the UUID mechanism could be implemented there. Maybe.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequester.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        client: Identifiable,
+        requestCode: Int
+    ): Observable<RequestPermissionsEvent>
+
+    data class CheckPermissionsResult(
+        val granted: List<String>,
+        val notGranted: List<String>,
+        val shouldShowRationale: List<String>
+    ) {
+        fun allGranted() : Boolean =
```
Comment (Nublo) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequester.kt]: Can we replace it to val, not fun? for me looks like a val as other fields

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Identifiable.kt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package com.badoo.ribs.core
+
+interface Identifiable {
+
+    fun id(): String
```
Comment (Nublo) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Identifiable.kt]: how about updating to val?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+import com.badoo.ribs.android.PermissionRequester.CheckPermissionsResult
+import com.badoo.ribs.android.PermissionRequester.RequestPermissionsEvent
+import com.badoo.ribs.android.PermissionRequester.RequestPermissionsEvent.Cancelled
+import com.badoo.ribs.android.PermissionRequester.RequestPermissionsEvent.RequestPermissionsResult
+import com.badoo.ribs.core.Identifiable
+import com.badoo.ribs.core.requestcode.RequestCodeRegistry
+import com.jakewharton.rxrelay2.PublishRelay
+import io.reactivex.Observable
+
+class PermissionRequesterImpl(
```
[the remark this answers is not in the document]
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: !! is to say know what you are doing. Internal design of the class guarantees that the call works (there's `ensureSubject()` for that), and if it doesn't, we should know about it asap.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequester.kt (hunk 3 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequester.kt]: In my mind it's a fun, but I'm open to hear others as well.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Identifiable.kt (hunk 4 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Identifiable.kt]: Any strong reasons why?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    private val requestCodeRegistry: RequestCodeRegistry
+) : PermissionRequester, PermissionRequestResultHandler {
+
+    private val events = HashMap<Int, HashMap<Int, PublishRelay<RequestPermissionsEvent>>>()
+
+    override fun checkPermissions(
+        client: Identifiable,
+        permissions: Array<String>
+    ) : CheckPermissionsResult {
+        val granted = mutableListOf<String>()
```
[the remark this answers is not in the document]
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: I don't see a problem with using a mutable structure as an intermediate step. Also this makes three passes on the input instead of just one. Not that it's a big issue, but still. I'm open to hearing others though.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        )
+    }
+
+    private fun onPermissionRequestFinished(
+        id: Int,
+        requestCode: Int,
+        permissions: Array<out String>,
+        grantResults: IntArray
+    ) {
+        val granted = mutableListOf<String>()
```
[the remark this answers is not in the document]
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: Same as above: is it worth it?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (ShikaSD) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: I think mutable list here is more efficient and readable.
However, consider simplifying the logic inside, right now it is a bit hard to follow, extracting *Compat calls to extension functions for example.

[the remark this answers is not in the document]
Comment (lukaville) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: I like @Nublo approach more because for me it's more readable

[the remark this answers is not in the document]
Comment (ShikaSD) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: One more thought: using when looks a bit better:
```
permissions.forEach { permission ->
    when {
        permission.granted() -> granted += permission
        permission.showRationale() -> shouldShowRationale += permission
        else -> canAsk += permission
    }
}
```

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+            code
+        )
+    }
+
+    // TODO Consider what is better: with requestCode here (separate streams) or without (client code branching on single stream)
+    override fun events(client: Identifiable, requestCode: Int): Observable<RequestPermissionsEvent> {
+        val id = requestCodeRegistry.generateGroupId(client.id())
+        ensureSubject(id, requestCode)
+
+        return events[id]!![requestCode]!!
```
Comment (lukaville) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: Use getValue(id) instead of !! because it will throw more meaningful error (`no such key ...` instead of null pointer exception) and for me, it looks nicer than `!!`

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequester.kt (hunk 3 above)
Comment (ShikaSD) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequester.kt]: Agree with val here.
Could be an extension as well.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        ActivityCompat.requestPermissions(activity,
+            permissions,
+            code
+        )
+    }
+
+    // TODO Consider what is better: with requestCode here (separate streams) or without (client code branching on single stream)
+    override fun events(client: Identifiable, requestCode: Int): Observable<RequestPermissionsEvent> {
+        val id = requestCodeRegistry.generateGroupId(client.id())
+        ensureSubject(id, requestCode)
```
Comment (lukaville) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: maybe return the newly created relay from `ensureSubject`?
so we can use non-null subject straightaway

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Identifiable.kt (hunk 4 above)
Comment (Nublo) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Identifiable.kt]: no strong reason just more nice API. You can keep it is as fun I am not insist.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 2 above)
Comment (lukaville) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]:
```suggestion
    private val events = HashMap<Int, HashMap<Int, Relay<RequestPermissionsEvent>>>()
```

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: I'll leave it as it is.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: I'll leave it as it is.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: This approach is the easiest to understand so far, I'll go with it, thanks.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequester.kt (hunk 3 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequester.kt]: Done.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Identifiable.kt (hunk 4 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Identifiable.kt]: Done.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: Done.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 8 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: Done.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 9 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: Done.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 2 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: Done.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Interactor.kt (hunk 1 above)
Comment (jeksor) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Interactor.kt]: What the solution?
Are you gonna remove `tag` from `Node` together?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequester.kt (hunk 3 above)
Comment (jeksor) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequester.kt]: `val` from my point of view.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Identifiable.kt (hunk 4 above)
Comment (jeksor) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Identifiable.kt]: **property** describes state.
**function** describes behaviour.

`id` is clearly not a behaviour, imho.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeksor) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: I like @Nublo suggestion, I find it more readable.
Three passes over the list of requested permission don't affect anyone, we can sacrifice that and increase readability.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (jeksor) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: We can apply the same idea proposed above by @Nublo and make the code a bit more readable as well.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 8 above)
Comment (jeksor) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: Agreed with @lukaville.
In `PermissionRequesterImpl` file, we can replace accessing map value with `getValue`.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 9 above)
Comment (jeksor) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: +1

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 2 above)
Comment (jeksor) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: I like the current approach with streams.
From the user perspective it's easier, looks nice. From our side: map of map, yes, it's not obvious. Let's try to keep API and make this map easier to understand.

Naming, comment, separate meaningful methods for operating over the map.

File: android/libraries/rib-android/src/main/java/com/uber/rib/core/RibActivity.kt [hunk 10]
```diff
@@ -6,9 +6,23 @@ import android.support.v7.app.AppCompatActivity
...
+import io.reactivex.Observable
 
-abstract class RibActivity : AppCompatActivity(), ActivityStarter, IntentCreator {
+abstract class RibActivity : AppCompatActivity(),
+    ActivityStarter,
+    IntentCreator,
+    PermissionRequester {
+
+    private val requestCodeRegistry = RequestCodeRegistry()
+    private val permissionRequester = PermissionRequesterImpl(
```
Comment (jeksor) [inline android/libraries/rib-android/src/main/java/com/uber/rib/core/RibActivity.kt]: Code style question.
Why are you braking lines almost for every `function` or `constructor` even if those contain 2, 3 arguments and fits the line (less then 150 symbols)?

This is just one of the examples.

I can't find such rule/recommendation either on [Kotlin Conventions](https://kotlinlang.org/docs/reference/coding-conventions.html) or [Android Kotlin Style Guide](https://developer.android.com/kotlin/style-guide).

cc @lukaville

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequester.kt [hunk 11]
```diff
@@ -0,0 +1,45 @@
+package com.badoo.ribs.android
+
+import com.badoo.ribs.core.Identifiable
+import io.reactivex.Observable
+
+interface PermissionRequester {
+
+    fun checkPermissions(
```
Comment (jeksor) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequester.kt]: Back to code style.
Here it's hard to distinguish `function`'s and `data class`es because of line breaking for functions.

It's quite understandable to do it for classes, but short functions. I don't think so.

File: android/libraries/rib-android/src/main/java/com/uber/rib/core/RibActivity.kt (hunk 10 above)
Comment (arkivanov) [reply android/libraries/rib-android/src/main/java/com/uber/rib/core/RibActivity.kt]: +1
And in case of line breaking and assignment operator, I think it would more readable if you break right after "=".
So either like this (if it fits one line):
`private val permissionRequester = PermissionRequesterImpl(this, requestCodeRegistry)`
or like this (if does not fit):
```
private val permissionRequester = 
    PermissionRequesterImpl(
        line1,
        line2,
        line3,
        line4,
    )
```

I think indentations are very important here, because last closing ")" belongs to PermissionRequesterImpl and not to "private val permissionRequester".

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+        if (grantResults.isEmpty()) {
+            onPermissionRequestCancelled(id, code)
+
+        } else {
+            onPermissionRequestFinished(id, code, permissions, grantResults)
+        }
+    }
+
+    private fun onPermissionRequestCancelled(id: Int, requestCode: Int) {
+        events[id]!![requestCode]!!.accept(
```
Comment (lukaville) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: getValue()?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+        permissions.forEachIndexed { index, permission ->
+            if (grantResults[index] == PackageManager.PERMISSION_GRANTED) {
+                granted.add(permission)
+            } else {
+                denied.add(permission)
+            }
+        }
+
+        events[id]!![requestCode]!!.accept(
```
Comment (lukaville) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: getValue()?

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+    ): Relay<RequestPermissionsEvent> {
+        var subjectJustCreated = false
+
+        if (!events.containsKey(id)) {
+            events[id] = hashMapOf()
+            subjectJustCreated = true
+        }
+
+        if (!events[id]!!.contains(requestCode)) {
+            events[id]!![requestCode] = PublishRelay.create()
```
Comment (lukaville) [inline android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: getValue()?

File: android/libraries/rib-android/src/main/java/com/uber/rib/core/RibActivity.kt (hunk 10 above)
Comment (zsoltk) [reply android/libraries/rib-android/src/main/java/com/uber/rib/core/RibActivity.kt]: Fixed.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 12 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: Fixed.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 13 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: Fixed.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 14 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: Fixed.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequester.kt (hunk 11 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequester.kt]: Fixed.

File: android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt (hunk 2 above)
Comment (zsoltk) [reply android/libraries/rib-base/src/main/java/com/badoo/ribs/android/PermissionRequesterImpl.kt]: Going with single stream approach after group discussion
